### PR TITLE
Issue/820/allow access to indexed variables

### PR DESCRIPTION
**Pull request by tschlat**
I have fixed the issue and added TestCases.

Please review.

### CODE REVIEW

File: core/citrus-api/src/main/java/com/consol/citrus/util/VariableExpressionIterator.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import com.consol.citrus.util.VariableExpressionIterator.VariableSegment;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+import java.lang.reflect.Array;
+import java.lang.reflect.Field;
+import java.util.Iterator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class VariableExpressionIterator implements Iterator<VariableSegment> {
```
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/util/VariableExpressionIterator.java:14]: this looks like a proper place for a short comment on how this iterator works, no? I'd be lost without knowing the issue and having access to the test class (`TestContextTest`).

File: core/citrus-api/src/main/java/com/consol/citrus/util/VariableExpressionIterator.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        public String getIndex() {
+            return index;
+        }
+
+        public Object getSegmentValue() {
+            return segmentValue;
+        }
+    }
+
+}
```
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/util/VariableExpressionIterator.java:179]: you could swap lines 178 and 179 - it's good pattern to have a newline at the end of the file.

File: core/citrus-base/src/test/java/com/consol/citrus/context/TestContextTest.java [hunk 3]
```diff
@@ -148,8 +164,10 @@ public void testGetVariableFromPathExpression() {
         Assert.assertEquals(context.getVariable("${container.data.data}"), "nested");
         Assert.assertEquals(context.getVariable("${container.data.number}"), "99");
         Assert.assertEquals(context.getVariable("${container.data.CONSTANT}"), "FOO");
+        Assert.assertEquals(context.getVariable("${container.intVals[1]}"), "1");
+        Assert.assertEquals(context.getVariable("${containerArray[3].data[1].data}"), "B");
     }
-
+
```
Comment (bbortt) [inline core/citrus-base/src/test/java/com/consol/citrus/context/TestContextTest.java:170]: remove those spaces, please.

File: core/citrus-base/src/test/java/com/consol/citrus/context/TestContextTest.java [hunk 4]
```diff
@@ -173,6 +191,13 @@ public void testUnknownFromPathExpression() {
         } catch (CitrusRuntimeException e) {
             Assert.assertEquals(e.getMessage(), "Unknown variable 'something.else'");
         }
+        
+        try {
+            context.getVariable("${helloData[1]}");
+            Assert.fail("Missing exception due to indexed access to non array variable");
+        } catch (CitrusRuntimeException e) {
+            Assert.assertTrue(e.getMessage().endsWith(""));
+        }
```
Comment (bbortt) [inline core/citrus-base/src/test/java/com/consol/citrus/context/TestContextTest.java:200]: `assertThrows` would probably be an option here.

**Review by bbortt — CHANGES_REQUESTED** [review]
wow, great job! that's a nice solution.. :) just one or two things to consider.

**Comment by bbortt** [comment]
closes #820.

File: core/citrus-base/src/test/java/com/consol/citrus/context/TestContextTest.java (hunk 4 above)
Comment (tschlat) [reply core/citrus-base/src/test/java/com/consol/citrus/context/TestContextTest.java:200]: Just following the pattern used in this TestCase.

**Review by bbortt — APPROVED** [review]
I've rebased it instead of merging `main` @tschlat 😉

/lgtm

**Comment by tschlat** [comment]
I close this PR as it is contained in https://github.com/citrusframework/citrus/pull/827

**Pull request closed without merging**